pkg/sdn/plugin/networkpolicy.go

```diff
@@ -0,0 +1,362 @@
+package plugin
+
+import (
+	"encoding/json"
+	"reflect"
+	"sort"
+	"sync"
+
+	"github.com/golang/glog"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kapierrs "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	"k8s.io/kubernetes/pkg/client/cache"
+	ktypes "k8s.io/kubernetes/pkg/types"
+	utilwait "k8s.io/kubernetes/pkg/util/wait"
+
+	osapi "github.com/openshift/origin/pkg/sdn/api"
+)
+
+type networkPolicyPlugin struct {
+	node  *OsdnNode
+	vnids *nodeVNIDMap
+
+	lock        sync.Mutex
+	namespaces  map[uint32]*npNamespace
+	kNamespaces map[string]kapi.Namespace
+}
+
+// npNamespace tracks NetworkPolicy-related data for a Namespace
+type npNamespace struct {
+	name     string
+	vnid     uint32
+	isolated bool
+	refs     int
+	inUse    bool
+
+	policies map[ktypes.UID]*npPolicy
+}
+
+// npPolicy is a parsed version of a single NetworkPolicy object
+type npPolicy struct {
+	policy extensions.NetworkPolicy
+
+	flows []string
+}
+
+func NewNetworkPolicyPlugin() osdnPolicy {
+	return &networkPolicyPlugin{
+		namespaces:  make(map[uint32]*npNamespace),
+		kNamespaces: make(map[string]kapi.Namespace),
+	}
+}
+
+func (np *networkPolicyPlugin) Name() string {
+	return osapi.NetworkPolicyPluginName
+}
+
+func (np *networkPolicyPlugin) Start(node *OsdnNode) error {
+	np.node = node
+	np.vnids = newNodeVNIDMap(np, node.osClient)
+	if err := np.vnids.Start(); err != nil {
+		return err
+	}
+
+	otx := node.ovs.NewTransaction()
+	otx.AddFlow("table=21, priority=100, ip, actions=ct(commit),goto_table:30")
+	otx.AddFlow("table=80, priority=50, ip, ct_state=-trk, actions=ct(commit,table=81)")
+	otx.AddFlow("table=81, priority=100, ip, ct_state=+trk+est, actions=output:NXM_NX_REG2[]")
+	otx.AddFlow("table=81, priority=0, actions=drop")
+	if err := otx.EndTransaction(); err != nil {
+		return err
+	}
+
+	if err := np.initNamespaces(); err != nil {
+		return err
+	}
+	if err := np.node.SetupEgressNetworkPolicy(); err != nil {
+		return err
+	}
+
+	go utilwait.Forever(np.watchNamespaces, 0)
+	go utilwait.Forever(np.watchNetworkPolicies, 0)
+	return nil
+}
+
+func (np *networkPolicyPlugin) initNamespaces() error {
+	np.lock.Lock()
+	defer np.lock.Unlock()
+
+	namespaces, err := np.node.kClient.Namespaces().List(kapi.ListOptions{})
```

Comment:
Yes. With the multitenant policy, everything it needs to track is on the NetNamespace object, but here we need to combine Namespace, NetNamespace, and NetworkPolicies.